File: Serialization/C++/Array/Array.cpp
#include "Array.h"

void Array::extendArray() {
    int newSize = this->size * 2;
    ArNode* newHead = new ArNode[newSize];
    
    for (int i = 0; i < this->len; i++) {
        newHead[i].data = this->head[i].data;
    }

    delete[] this->head;
    this->head = newHead;
    this->size = newSize;
}

Array::Array(int size) : len(0)
                       , size(size) {
    if (size < 1) { 
        throw runtime_error("Невозможно создать массив нулевого размера"); 
    }
    this->head = new ArNode[size];
}

Array::Array(initializer_list<string> list) : len(0)
                                            , size(list.size()) {
    this->head = new ArNode[size];
    for (string item : list) {
        addElementEnd(item);
    }
}

Array::~Array() {
    delete[] this->head;
    this->head = nullptr;  
    this->len = 0;
    this->size = 0;
}

string Array::getElement(int index) {
    if (index < 0 || index >= this->len) { 
        throw range_error("Выход за пределы массива"); 
    }
    return this->head[index].data;
}

void Array::setElement(string key, int index) {
    if (index < 0 || index >= this->len) { 
        throw range_error("Выход за пределы массива"); 
    }
    this->head[index].data = key;
}

void Array::deleteElement(int index) {
    if (index < 0 || index >= this->len) { 
        throw range_error("Выход за пределы массива"); 
    }
    for (int i = index; i < this->len - 1; i++) {
        this->head[i].data = this->head[i + 1].data;  
    }
    this->len -= 1;
}

void Array::addElementAtIndex(string key, int index) {
    if (index < 0 || index > this->len) { 
        throw range_error("Выход за пределы массива"); 
    }
    if (this->len >= this->size) { 
        extendArray(); 
    }
    for (int i = this->len; i > index; i--) {
        this->head[i].data = this->head[i - 1].data;  
    }
    this->head[index].data = key;
    this->len += 1;
}

void Array::addElementEnd(string key) {
    if (this->len >= this->size) {
        extendArray(); 
    }

    this->head[this->len].data = key;
    this->len += 1;
}

int Array::getLength() const {
    return this->len;
}

ArNode* Array::getHead() {
    return this->head;
}

int Array::getSize() const {
    return this->size;
}

int Array::isInArray(const string& key) const {
    for (int i = 0; i < this->len; i++) {
        if (this->head[i].data == key) {
            return i;
        }
    }
    return -1;
}

void Array::print() {
    for (size_t i = 0; i < this->len - 1; i++) {
        cout << this->head[i].data << " ";
    }
    cout << this->head[this->len - 1].data << endl;
}

void Array::writeBinary(const string& filename) {
    ofstream file(filename, ios::binary);
    if (!file) {
        throw runtime_error("Не удалось открыть файл для записи: " + filename);
    }

    file.write(reinterpret_cast<const char*>(&this->len), sizeof(this->len));

    for (size_t i = 0; i < this->len; i++) {
        size_t stringLength = this->head[i].data.length();
        file.write(reinterpret_cast<const char*>(&stringLength), sizeof(stringLength));
        file.write(this->head[i].data.c_str(), stringLength);
    }
    
    file.close();
}

void Array::readBinary(const string& filename){
    ifstream file(filename, ios::binary);
    if (!file) {
        throw runtime_error("Не удалось открыть файл: " + filename);
    }

    file.read(reinterpret_cast<char*>(&this->size), sizeof(this->size));
    delete[] head;
    this->head = new ArNode[this->size];

    for (int i = 0; i < this->size; i++) {
        size_t stringLength;
        file.read(reinterpret_cast<char*>(&stringLength), sizeof(stringLength));
        
        string inputString;
        inputString.resize(stringLength);
        file.read(inputString.data(), stringLength);
        
        this->head[i].data = inputString;
        this->len++;
    }

    file.close();
}

void Array::writeText(const string& filename) {
    ofstream file(filename);
    if (!file) {
        throw runtime_error("Не удалось открыть файл для записи: " + filename);
    }

    file << this->len << endl;

    for (size_t i = 0; i < this->len; i++) {
        file << this->head[i].data << endl;
    }
    
    file.close();
}

void Array::readText(const string& filename){
    ifstream file(filename);
    if (!file) {
        throw runtime_error("Не удалось открыть файл: " + filename);
    }

    string inputSize;
    getline(file, inputSize);
    this->size = stoi(inputSize);
    this->head = new ArNode[this->size];

    for (int i = 0; i < this->size; i++) {
        string inputString;
        getline(file, inputString);
        
        this->head[i].data = inputString;
        this->len++;
    }

    file.close();
}

File: Tests/Go/array/array.go
package array

import (
	"bufio"
	"encoding/binary"
	"errors"
	"fmt"
	"os"
	"strconv"
)

type Array struct {
	data []string 
	len  int      
	cap  int      
}

func NewArray(size int) (*Array, error) {
	if size < 1 {
		return nil, errors.New("cannot create array of zero size")
	}
	return &Array{
		data: make([]string, size),
		len:  0,
		cap:  size,
	}, nil
}

func NewArrayFromList(items []string) (*Array, error) {
	if len(items) < 1 {
		return NewArray(1)
	}
	a, _ := NewArray(len(items))
	for _, item := range items {
		a.AddElementEnd(item)
	}
	return a, nil
}

func (a *Array) grow() {
	newCap := a.cap * 2
	newData := make([]string, newCap)
	copy(newData, a.data[:a.len])
	a.data = newData
	a.cap = newCap
}

func (a *Array) GetElement(index int) (string, error) {
	if index < 0 || index >= a.len {
		return "", errors.New("index out of bounds")
	}
	return a.data[index], nil
}

func (a *Array) SetElement(key string, index int) error {
	if index < 0 || index >= a.len {
		return errors.New("index out of bounds")
	}
	a.data[index] = key
	return nil
}

func (a *Array) DeleteElement(index int) error {
	if index < 0 || index >= a.len {
		return errors.New("index out of bounds")
	}
	for i := index; i < a.len-1; i++ {
		a.data[i] = a.data[i+1]
	}
	a.len--
	return nil
}

func (a *Array) AddElementAtIndex(key string, index int) error {
	if index < 0 || index > a.len {
		return errors.New("index out of bounds")
	}
	if a.len >= a.cap {
		a.grow()
	}
	for i := a.len; i > index; i-- {
		a.data[i] = a.data[i-1]
	}
	a.data[index] = key
	a.len++
	return nil
}

func (a *Array) AddElementEnd(key string) {
	if a.len >= a.cap {
		a.grow()
	}
	a.data[a.len] = key
	a.len++
}

func (a *Array) GetLength() int {
	return a.len
}

func (a *Array) GetCapacity() int {
	return a.cap
}

func (a *Array) IsInArray(key string) int {
	for i := 0; i < a.len; i++ {
		if a.data[i] == key {
			return i
		}
	}
	return -1
}

func (a *Array) Print() {
	for i := 0; i < a.len; i++ {
		if i > 0 {
			fmt.Print(" ")
		}
		fmt.Print(a.data[i])
	}
	fmt.Println()
}

func (a *Array) WriteBinary(filename string) error {
	file, err := os.Create(filename)
	if err != nil {
		return fmt.Errorf("failed to open file for writing: %w", err)
	}
	defer file.Close()

	if err := binary.Write(file, binary.LittleEndian, uint32(a.len)); err != nil {
		return err
	}

	for i := 0; i < a.len; i++ {
		s := a.data[i]
		length := uint32(len(s))
		if err := binary.Write(file, binary.LittleEndian, length); err != nil {
			return err
		}
		if _, err := file.Write([]byte(s)); err != nil {
			return err
		}
	}
	return nil
}

func (a *Array) ReadBinary(filename string) error {
	file, err := os.Open(filename)
	if err != nil {
		return fmt.Errorf("failed to open file: %w", err)
	}
	defer file.Close()

	var length uint32
	if err := binary.Read(file, binary.LittleEndian, &length); err != nil {
		return err
	}

	newCap := int(length)
	if newCap == 0 {
		newCap = 1
	}
	a.data = make([]string, newCap)
	a.cap = newCap
	a.len = 0

	for i := uint32(0); i < length; i++ {
		var strLen uint32
		if err := binary.Read(file, binary.LittleEndian, &strLen); err != nil {
			return err
		}
		buf := make([]byte, strLen)
		if _, err := file.Read(buf); err != nil {
			return err
		}
		a.data[i] = string(buf)
		a.len++
	}
	return nil
}

func (a *Array) WriteText(filename string) error {
	file, err := os.Create(filename)
	if err != nil {
		return fmt.Errorf("failed to open file for writing: %w", err)
	}
	defer file.Close()

	writer := bufio.NewWriter(file)
	defer writer.Flush()

	if _, err := fmt.Fprintf(writer, "%d\n", a.len); err != nil {
		return err
	}

	for i := 0; i < a.len; i++ {
		if _, err := fmt.Fprintln(writer, a.data[i]); err != nil {
			return err
		}
	}
	return nil
}

func (a *Array) ReadText(filename string) error {
	file, err := os.Open(filename)
	if err != nil {
		return fmt.Errorf("failed to open file: %w", err)
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	if !scanner.Scan() {
		return errors.New("empty file")
	}
	lengthStr := scanner.Text()
	length, err := strconv.Atoi(lengthStr)
	if err != nil {
		return fmt.Errorf("invalid length line: %w", err)
	}

	newCap := length
	if newCap == 0 {
		newCap = 1
	}
	a.data = make([]string, newCap)
	a.cap = newCap
	a.len = 0

	for i := 0; i < length; i++ {
		if !scanner.Scan() {
			return errors.New("unexpected EOF")
		}
		a.data[i] = scanner.Text()
		a.len++
	}
	return nil
}